### PR TITLE
*: add max1row info and refine scalar subquery's decorrelation.

### PR DESCRIPTION
**Pull request by hanfei1991**
1. add max1row info for schema structure.
2. For scalar subquery, we should use inner join instead of left outer join.
3. implement left outer join for nested loop join.
@coocood @shenli @zimulala PTAL

### CODE REVIEW

**Comment by shenli** [comment]
@hanfei1991 CI failed

File: plan/physical_plan_test.go [hunk 1]
```diff
@@ -765,7 +765,7 @@ func (s *testPlanSuite) TestAddCache(c *C) {
 	}{
 		{
 			sql: "select * from t t1 where t1.a=(select min(t2.a) from t t2, t t3 where t2.a=t3.a and t2.b > t1.b + t3.b)",
-			ans: "Apply{Table(t)->LeftHashJoin{Table(t)->Cache->Table(t)->Cache}(t2.a,t3.a)->StreamAgg->MaxOneRow}->Projection",
+			ans: "Apply{Table(t)->LeftHashJoin{Table(t)->Cache->Table(t)->Cache}(t2.a,t3.a)->StreamAgg->MaxOneRow}->Selection->Projection",
```
Comment (coocood) [inline plan/physical_plan_test.go]: Why the extra `Selection`?

File: plan/decorrelate.go [hunk 2]
```diff
@@ -100,7 +109,7 @@ func (s *decorrelateSolver) optimize(p LogicalPlan, _ context.Context, _ *idAllo
 	for _, child := range p.Children() {
 		np, _ := s.optimize(child.(LogicalPlan), nil, nil)
 		newChildren = append(newChildren, np)
-		child.SetParents(p)
+		np.SetParents(p)
```
Comment (coocood) [inline plan/decorrelate.go]: Is this a bug?
Any test cover it?

Comment (hanfei1991) [inline plan/decorrelate.go]: This is a bug after I begin to process the case of proj.

File: plan/physical_plan_test.go (hunk 1 above)
Comment (hanfei1991) [inline plan/physical_plan_test.go]: Because the jointype of Apply has been changed to LeftOuterJoin,

File: plan/decorrelate.go (hunk 2 above)
Comment (coocood) [inline plan/decorrelate.go]: Please add a test case cover it.

Comment (hanfei1991) [inline plan/decorrelate.go]: the existant test can cover this.

Comment (coocood) [inline plan/decorrelate.go]: I change it back to `child`, the test in plan package still pass.

Comment (hanfei1991) [inline plan/decorrelate.go]: executor test will fail

**Comment by coocood** [comment]
LGTM

**Comment by hanfei1991** [comment]
@zimulala PTAL

File: plan/build_key_info.go [hunk 3]
```diff
@@ -39,6 +40,30 @@ func (p *Aggregation) buildKeyInfo() {
...
+		return false
+	}
+	return p.children[0].Schema().IsUniqueKey(col)
+}
+
+func (p *Selection) buildKeyInfo() {
+	p.baseLogicalPlan.buildKeyInfo()
+	p.schema.MaxOneRow = p.children[0].Schema().MaxOneRow
+	for _, cond := range p.Conditions {
+		if sf, ok := cond.(*expression.ScalarFunction); ok && sf.FuncName.L == ast.EQ {
```
Comment (zimulala) [inline plan/build_key_info.go]: Add a condition about constant.

Comment (hanfei1991) [inline plan/build_key_info.go]: @zimulala PTAL

**Comment by zimulala** [comment]
LGTM

**Pull request merged**